File: src/pgo/model/tla/TLAInstance.java
package pgo.model.tla;

import java.util.List;
import java.util.stream.Collectors;

import pgo.util.SourceLocation;

/**
 * 
 * TLA AST PlusCalNode:
 * 
 * INSTANCE ModuleName (WITH a <- <expr>, b <- <expr>)?
 *
 */
public class TLAInstance extends TLAUnit {
	private TLAIdentifier moduleName;
	private List<Remapping> remappings;
	private boolean local;
	
	public TLAInstance(SourceLocation location, TLAIdentifier moduleName, List<Remapping> remappings, boolean isLocal) {
		super(location);
		this.moduleName = moduleName;
		this.remappings = remappings;
		this.local = isLocal;
	}
	
	@Override
	public TLAInstance copy() {
		return new TLAInstance(getLocation(), moduleName.copy(), remappings.stream().map(Remapping::copy).collect(Collectors.toList()), local);
	}
	
	public static class Remapping extends TLANode {
		TLAIdentifier from;
		TLAExpression to;
		public Remapping(SourceLocation location, TLAIdentifier from, TLAExpression to) {
			super(location);
			this.from = from;
			this.to = to;
		}
		public TLAIdentifier getFrom() {
			return from;
		}
		public TLAExpression getTo() {
			return to;
		}
		@Override
		public Remapping copy() {
			return new Remapping(getLocation(), from.copy(), to.copy());
		}
		@Override
		public <T, E extends Throwable> T accept(TLANodeVisitor<T, E> v) throws E {
			return v.visit(this);
		}
		@Override
		public int hashCode() {
			final int prime = 31;
			int result = 1;
			result = prime * result + ((from == null) ? 0 : from.hashCode());
			result = prime * result + ((to == null) ? 0 : to.hashCode());
			return result;
		}
		@Override
		public boolean equals(Object obj) {
			if (this == obj)
				return true;
			if (obj == null)
				return false;
			if (getClass() != obj.getClass())
				return false;
			Remapping other = (Remapping) obj;
			if (from == null) {
				if (other.from != null)
					return false;
			} else if (!from.equals(other.from))
				return false;
			if (to == null) {
				return other.to == null;
			} else return to.equals(other.to);
		}
	}
	
	public TLAIdentifier getModuleName() {
		return moduleName;
	}
	
	public List<Remapping> getRemappings(){
		return remappings;
	}
	
	public boolean isLocal() {
		return local;
	}
	
	@Override
	public <T, E extends Throwable> T accept(TLAUnitVisitor<T, E> v) throws E {
		return v.visit(this);
	}

	@Override
	public int hashCode() {
		final int prime = 31;
		int result = 1;
		result = prime * result + (local ? 1231 : 1237);
		result = prime * result + ((moduleName == null) ? 0 : moduleName.hashCode());
		result = prime * result + ((remappings == null) ? 0 : remappings.hashCode());
		return result;
	}

	@Override
	public boolean equals(Object obj) {
		if (this == obj)
			return true;
		if (obj == null)
			return false;
		if (getClass() != obj.getClass())
			return false;
		TLAInstance other = (TLAInstance) obj;
		if (local != other.local)
			return false;
		if (moduleName == null) {
			if (other.moduleName != null)
				return false;
		} else if (!moduleName.equals(other.moduleName))
			return false;
		if (remappings == null) {
			return other.remappings == null;
		} else return remappings.equals(other.remappings);
	}

}

File: src/pgo/model/tla/TLAInstance.scala
package pgo.model.tla

import pgo.util.SourceLocation

import scala.jdk.CollectionConverters._

final class TLAInstance(loc: SourceLocation, val moduleName: TLAIdentifier, val remappings: List[TLAInstance.Remapping],
                  val isLocal: Boolean) extends TLAUnit(loc) with TLADefinitionComposite {
  var module: Option[TLAModule] = None

  def this(loc: SourceLocation, moduleName: TLAIdentifier, remappings: java.util.List[TLAInstance.Remapping], isLocal: Boolean) =
    this(loc, moduleName, remappings.asScala.toList, isLocal)

  override val definitions: List[TLADefinition] = List(this)

  def getModuleName: TLAIdentifier = moduleName
  def getRemappings: java.util.List[TLAInstance.Remapping] = remappings.asJava

  override def copy: TLAInstance =
    new TLAInstance(loc, moduleName.copy(), remappings.map(_.copy()), isLocal)

  override def members: List[TLADefinition] =
    module.get.definitions

  override def accept[T, E <: Throwable](v: TLAUnitVisitor[T, E]): T = v.visit(this)

  def canEqual(other: Any): Boolean = other.isInstanceOf[TLAInstance]

  override def equals(other: Any): Boolean = other match {
    case that: TLAInstance =>
      (that canEqual this) &&
        moduleName == that.moduleName &&
        remappings == that.remappings &&
        isLocal == that.isLocal
    case _ => false
  }

  override def hashCode(): Int = {
    val state = Seq(moduleName, remappings, isLocal)
    state.map(_.hashCode()).foldLeft(0)((a, b) => 31 * a + b)
  }
}

object TLAInstance {
  final class Remapping(loc: SourceLocation, val from: TLAIdentifier, val to: TLAExpression) extends TLANode(loc) with TLADefinitionOne {
    override def copy(): Remapping =
      new Remapping(loc, from.copy(), to.copy())

    def getFrom: TLAIdentifier = from
    def getTo: TLAExpression = to

    override def accept[T, E <: Throwable](v: TLANodeVisitor[T, E]): T = v.visit(this)

    override def arity: Int = 0

    override def isModuleInstance: Boolean = false

    override def identifier: TLAIdentifier = from

    override def scope: Map[TLAIdentifier, TLADefinitionOne] = Map.empty

    override def equals(other: Any): Boolean = other match {
      case that: Remapping =>
        from == that.from &&
          to == that.to
      case _ => false
    }

    override def hashCode(): Int = {
      val state = Seq(from, to)
      state.map(_.hashCode()).foldLeft(0)((a, b) => 31 * a + b)
    }
  }
}
